File: bestia-client/src/Bnet/Message/Entity/VanishEntitySMSG.cs
using Bnet;
using Godot;

namespace BestiaBehemothClient.Bnet.Message.Entity
{
  /// <summary>
  /// An entity the client should stop drawing, and why.
  /// <para>
  /// The kind is exposed as predicates rather than as a property because GDScript cannot read the proto
  /// enum. <see cref="IsOutOfSight"/> is the one that must not play a send-off: it means the entity is
  /// alive and simply outside this client's view, and it can come back within the same second.
  /// </para>
  /// </summary>
  [GlobalClass]
  public partial class VanishEntitySMSG : EntitySMSG
  {
    private VanishKind Kind { get; set; } = VanishKind.Gone;

    public bool IsDead()
    {
      return Kind == VanishKind.Death;
    }

    public bool IsGone()
    {
      return Kind == VanishKind.Gone;
    }

    /// <summary>Left the client's view while still alive - the chunk it stands in is no longer held.</summary>
    public bool IsOutOfSight()
    {
      return Kind == VanishKind.OutOfSight;
    }

    public static VanishEntitySMSG FromProto(global::Bnet.VanishEntitySMSG protoVanish)
    {
      return new VanishEntitySMSG
      {
        EntityId = protoVanish.EntityId,
        Kind = protoVanish.Kind
      };
    }
  }
}

File: zone-server/src/main/kotlin/net/bestia/zone/entity/VanishEntitySMSG.kt
package net.bestia.zone.entity

import net.bestia.bnet.proto.EnvelopeProto
import net.bestia.bnet.proto.VanishEntitySmsgProto
import net.bestia.zone.message.EntitySMSG
import net.bestia.zone.util.EntityId

data class VanishEntitySMSG(
  override val entityId: EntityId,
  val kind: VanishKind
) : EntitySMSG {

  enum class VanishKind {
    /** Removed from the world for a reason the client is not told. */
    GONE,

    /** Killed, so a death animation is worth playing. */
    DEATH,

    /**
     * Still alive, just no longer in view - the chunk it stands in has left this client's subscription.
     *
     * Distinct from [GONE] because it must not play a send-off: an entity pacing a chunk boundary would
     * otherwise fade out and back repeatedly, and the client has to be free to drop the node at once.
     */
    OUT_OF_SIGHT
  }

  override fun toBnetEnvelope(): EnvelopeProto.Envelope {
    val kind = when (kind) {
      VanishKind.GONE -> VanishEntitySmsgProto.VanishKind.GONE
      VanishKind.DEATH -> VanishEntitySmsgProto.VanishKind.DEATH
      VanishKind.OUT_OF_SIGHT -> VanishEntitySmsgProto.VanishKind.OUT_OF_SIGHT
    }

    val vanishMsg = VanishEntitySmsgProto.VanishEntitySMSG.newBuilder()
      .setEntityId(entityId)
      .setKind(kind)

    return EnvelopeProto.Envelope.newBuilder()
      .setVanishEntity(vanishMsg)
      .build()
  }
}